File: src/main/java/io/beaniejoy/dongnecafe/cafe/domain/CafeMenu.java
package io.beaniejoy.dongnecafe.cafe.domain;

import io.beaniejoy.dongnecafe.common.domain.BaseTimeEntity;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.GenericGenerator;

import javax.persistence.*;
import java.util.List;
import java.util.UUID;

@Getter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
public class CafeMenu extends BaseTimeEntity {

    @Id
    @GeneratedValue(generator = "uuid2")
    @GenericGenerator(name = "uuid2", strategy = "uuid2")
    @Column(name = "menu_id", columnDefinition = "BINARY(16)")
    private UUID menuId;

    private String name;

    private Integer price;

    @ManyToOne
    @JoinColumn(name = "cafe_id")
    private Cafe cafe;

    @OneToMany(mappedBy = "cafeMenu")
    private List<MenuOption> menuOptionList;
}

File: src/main/java/io/beaniejoy/dongnecafe/domain/cafe/entity/CafeMenu.kt
package io.beaniejoy.dongnecafe.domain.cafe.entity

import io.beaniejoy.dongnecafe.common.entity.BaseTimeEntity
import io.beaniejoy.dongnecafe.domain.cafe.dto.request.MenuOptionInfoRequestDto
import java.math.BigDecimal
import javax.persistence.*

@Entity
@Table(name = "cafe_menu")
class CafeMenu protected constructor(
    name: String,
    price: BigDecimal,
) : BaseTimeEntity() {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "cafe_menu_id", nullable = false)
    val id: Long = 0L

    @Column(name = "name", nullable = false)
    val name: String = name

    @Column(name = "price", nullable = false)
    val price: BigDecimal = price

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "cafe_id", nullable = false)
    var cafe: Cafe? = null
        protected set

    @OneToMany(mappedBy = "cafeMenu", fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
    val menuOptionList: MutableList<MenuOption> = arrayListOf()

    companion object {
        fun createCafeMenu(name: String, price: BigDecimal, menuOptionRequestList: List<MenuOptionInfoRequestDto>): CafeMenu {
            val menuOptionEntityList = menuOptionRequestList.map { menuOptionRequestDto ->
                MenuOption.createMenuOption(
                    title = menuOptionRequestDto.title,
                    optionDetailRequestList = menuOptionRequestDto.optionDetailList
                )
            }

            return CafeMenu(
                name = name,
                price = price
            ).apply {
                menuOptionEntityList.forEach { this.addMenuOption(it) }
            }
        }
    }

    fun updateCafe(cafe: Cafe) {
        this.cafe = cafe
    }

    fun addMenuOption(menuOption: MenuOption) {
        this.menuOptionList.add(menuOption)
        menuOption.updateCafeMenu(this)
    }
}